crates/router/src/types/handler.rs

```diff
@@ -0,0 +1,125 @@
+pub use api_models::{
+    enums::{Connector, PayoutConnectors},
+    payouts as payout_types,
+};
+pub use common_utils::request::RequestBody;
+use data_models::payments::{payment_attempt::PaymentAttempt, payment_intent::PaymentIntent};
+use diesel_models::enums;
+
+use crate::{
+    core::{errors::RouterResult, payments::helpers},
+    routes::AppState,
+    types::api::{self, payments},
+};
+
+pub struct Oss;
+
+#[async_trait::async_trait]
+pub trait PaymentMethodRetrieve {
+    async fn retrieve_payment_method(
+        pm_data: &Option<payments::PaymentMethodData>,
+        state: &AppState,
+        payment_intent: &PaymentIntent,
+        payment_attempt: &PaymentAttempt,
+        payment_token: &mut Option<String>,
+    ) -> RouterResult<Option<payments::PaymentMethodData>>;
+}
+
+#[async_trait::async_trait]
+impl PaymentMethodRetrieve for Oss {
+    async fn retrieve_payment_method(
+        pm_data: &Option<payments::PaymentMethodData>,
+        state: &AppState,
+        payment_intent: &PaymentIntent,
+        payment_attempt: &PaymentAttempt,
+        payment_token: &mut Option<String>,
+    ) -> RouterResult<Option<payments::PaymentMethodData>> {
+        match pm_data {
+            pm_opt @ Some(pm @ api::PaymentMethodData::Card(_)) => {
+                if helpers::should_store_payment_method_data_in_vault(
+                    &state.conf.temp_locker_disable_config,
+                    payment_attempt.connector.clone(),
+                    enums::PaymentMethod::Card,
+                ) {
+                    let parent_payment_method_token = helpers::store_in_vault_and_generate_ppmt(
+                        state,
+                        pm,
+                        payment_intent,
+                        payment_attempt,
+                        enums::PaymentMethod::Card,
+                    )
+                    .await?;
+
+                    *payment_token = Some(parent_payment_method_token);
```

Comment:
Return the token from the function alongside the `PaymentMethodData` instead of mutating in-place, and set it in `make_pm_data`.  That would remove the need for the last function argument.